firedrake/matrix.py

```diff
@@ -1,30 +1,41 @@
 import abc
 import itertools
+import ufl
 
 from pyop2 import op2
 from pyop2.utils import as_tuple
 from firedrake.petsc import PETSc
+from types import SimpleNamespace
 
 
-class MatrixBase(object, metaclass=abc.ABCMeta):
+class MatrixBase(ufl.Matrix, metaclass=abc.ABCMeta):
     """A representation of the linear operator associated with a
     bilinear form and bcs.  Explicitly assembled matrices and matrix-free
     matrix classes will derive from this
 
-    :arg a: the bilinear form this :class:`MatrixBase` represents.
+    :arg a: the bilinear form this :class:`MatrixBase` represents
+            or a tuple of the arguments it represents
 
     :arg bcs: an iterable of boundary conditions to apply to this
         :class:`MatrixBase`.  May be `None` if there are no boundary
         conditions to apply.
     :arg mat_type: matrix type of assembled matrix, or 'matfree' for matrix-free
     """
     def __init__(self, a, bcs, mat_type):
```

Comment:
The base class should ideally not have to dispatch on types of arguments. I think now it is the case that every `MatrixBase` "has-a" arguments property, but only `Matrix` and `ImplicitMatrix` "has-a" `.a` form.

So I think the better refactoring is to make this:

```python
def __init__(self, test, trial, bcs, mat_type):
    self._arguments = test, trial
```

And then `Matrix` does

```
def __init__(self, a, ...):
    super().__init__(*a.arguments(), ...)
    self.a = a
```

while `AssembledMatrix` does
```
def __init__(self, test, trial, ...):
    super().__init__(test, trial, ...)
```